Userland/Libraries/LibGfx/DDSLoader.cpp

```diff
@@ -0,0 +1,1043 @@
+    builder.append("\n");
+
+    dbgln("{}", builder.to_string());
+}
+
+static RefPtr<Gfx::Bitmap> load_dds_impl(const u8* data, size_t length)
+{
+    DDSLoadingContext context;
+    context.data = data;
+    context.data_size = length;
+
+    if (!decode_dds(context))
+        return nullptr;
+
+    return context.bitmap;
+}
+
+RefPtr<Gfx::Bitmap> load_dds(String const& path)
+{
+    auto file_or_error = MappedFile::map(path);
+    if (file_or_error.is_error())
+        return nullptr;
+    auto bitmap = load_dds_impl((const u8*)file_or_error.value()->data(), file_or_error.value()->size());
+    if (bitmap)
+        bitmap->set_mmap_name(String::formatted("Gfx::Bitmap [{}] - Decoded DDS: {}", bitmap->size(), LexicalPath::canonicalized_path(path)));
+    return bitmap;
+}
+
+RefPtr<Gfx::Bitmap> load_dds_from_memory(const u8* data, size_t length)
+{
+    auto bitmap = load_dds_impl(data, length);
+    if (bitmap)
+        bitmap->set_mmap_name(String::formatted("Gfx::Bitmap [{}] - Decoded DDS: <memory>", bitmap->size()));
+    return bitmap;
+}
+
+DDSImageDecoderPlugin::DDSImageDecoderPlugin(const u8* data, size_t size)
+{
+    m_context = make<DDSLoadingContext>();
+    m_context->data = data;
+    m_context->data_size = size;
+}
+
+DDSImageDecoderPlugin::~DDSImageDecoderPlugin()
+{
+}
+
+IntSize DDSImageDecoderPlugin::size()
+{
+    if (m_context->state == DDSLoadingContext::State::Error)
+        return {};
+
+    if (m_context->state == DDSLoadingContext::State::BitmapDecoded)
+        return { m_context->header.width, m_context->header.height };
+
+    return {};
+}
+
+RefPtr<Gfx::Bitmap> DDSImageDecoderPlugin::bitmap()
+{
+    if (m_context->state == DDSLoadingContext::State::Error)
+        return nullptr;
+
+    if (m_context->state < DDSLoadingContext::State::BitmapDecoded) {
+        bool success = decode_dds(*m_context);
+        if (!success)
+            return nullptr;
+    }
+
+    VERIFY(m_context->bitmap);
+    return m_context->bitmap;
+}
+
+void DDSImageDecoderPlugin::set_volatile()
+{
+    if (m_context->bitmap)
+        m_context->bitmap->set_volatile();
+}
+
+bool DDSImageDecoderPlugin::set_nonvolatile()
+{
+    if (!m_context->bitmap)
+        return false;
+    return m_context->bitmap->set_nonvolatile();
+}
+
+bool DDSImageDecoderPlugin::sniff()
+{
+    // The header is always atleast 128 bytes, so if the file is smaller, it cant be a DDS.
+    return m_context->data_size > 128
+        && m_context->data[0] == 0x44
+        && m_context->data[1] == 0x44
+        && m_context->data[2] == 0x53
+        && m_context->data[3] == 0x20;
+}
+
+bool DDSImageDecoderPlugin::is_animated()
+{
+    return false;
+}
+
+size_t DDSImageDecoderPlugin::loop_count()
+{
+    return 0;
+}
+
+size_t DDSImageDecoderPlugin::frame_count()
+{
+    return 1;
+}
+
+ImageFrameDescriptor DDSImageDecoderPlugin::frame(size_t i)
+{
+    // We have "frames", but they are all the same image, so lets just use the largest version
+    (void)i;
```

Comment:
Please don't `(void)` variables, use `[[maybe_unused]] i` for the parameter instead.

Reply:
Or even better, you can simply omit the `i` parameter name and the compiler won't complain.

Reply:
No, in this case I think omitting the name makes it less clear - `i` makes it obvious that it's an index. Just `frame(size_t)` could be anything.

Reply:
hmm.. my line of thought was that `i` wasn't very descriptive, but I think you're right as `i` would definitely mean index!